### ISSUE TITLE
ShellCheck warning SC2320 triggered incorrectly in function arguments

### ISSUE THREAD

**Issue by TaurusOmar** [koalaman/shellcheck#2991]
#### For bugs
- Rule Id (if any, e.g. SC1000):  0.10.0
- My shellcheck version (`shellcheck --version` or "online"): 
- [ ] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ ] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

## Description 

When I use the `demo` function to print a list of arguments in a bash script, ShellCheck incorrectly issues warning SC2320. The warning suggests that `$?` refers to `echo` or `printf`, but in this case, it refers to the return code of the last command execution. This occurs on the line where the argument list is attempted to be printed with `echo "argument list: $@"`.

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/env bash

demo() {
    echo "in function demo..."
    echo "the script name: $0"
    echo "argument count (not include the script itself): $#"
    echo "the first argument: $1"
    echo "the second argument: $2"
    echo "the 10th argument: ${10}"

    ret=$?
    echo "command exited with ${ret}"
    if [[ ${ret} -ne 0 ]]; then
        echo "failed"
    fi
}

# We assign the list of arguments to a variable before printing
args_list="$@"
echo "argument list: $args_list"

demo 1 2 3

```

In this case, `$@` is assigned to `args_list` before printing, so `$?` will not be affected by the SC2320 warning and the argument list will be printed correctly.

**Comment by brother** [comment]
Duplicate #2990
